### PR TITLE
[dhcp6] implement DHCPv6 Prefix Delegation (PD) client

### PR DESCRIPTION
**Pull request by abtink**
This commit introduces the `Dhcp6PdClient` class, which implements DHCPv6 Prefix Delegation (PD) client functionality. It integrates with `BorderRouter::RoutingManager` and its `PdPrefixManager` sub-component. The CMake `OT_BORDER_ROUTING_DHCP6_PD_CLIENT` mapped to `OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_CLIENT_ENABLE` config enables this feature.

Previously, the platform layer was expected to implement client functionality, acquiring and providing the delegated prefix(es) to the OT stack using `otPlatBorderRouter*` callbacks. This approach continues to be supported. The `Dhcp6PdClient` feature adds native support for this functionality in the OpenThread core.

The `Dhcp6PdClient` implementation follows RFC 8415, focusing on prefix delegation and all required behaviors. The client follows the standard four-message Solicit/Advertise/Request/Reply exchange to obtain a delegated prefix, followed by a two-message Renew/Reply or Rebind/Reply exchange to extend the lifetime of the delegated prefix. When the prefix is no longer needed, a two-message Release/Reply exchange ends its lease. The current client implementation does not support the optional "Reconfigure Accept" mechanism.

A set of `otPlatInfraIfDhcp6PdClient*` platform APIs are also introduced for use by the `Dhcp6PdClient`. These APIs are used to enable or disable listening for DHCPv6 messages and to handle sending and receiving them on the standard client and server UDP ports (546 and 547), effectively acting as a UDP socket.

This commit also includes a comprehensive unit test covering various aspects of `Dhcp6PdClient`, including common behaviors and many specific edge cases.

### CODE REVIEW

File: src/core/border_router/dhcp6_pd_client.cpp [hunk 1]
```diff
@@ -0,0 +1,1301 @@
...
+    }
+#endif
+}
+
+void Dhcp6PdClient::RetxTracker::UpdateTimeoutAndCountAfterTx(void)
+{
+    // RFC 8415 Section 15 required retransmission timeout (RT) to be updated
+    // for each subsequent tx based on the previous value using `RT =
+    // 2*RTprev + RAND*RTprev` This is interpreted as `RT = RTPrev +
+    // (RTprev + RAND*RTprev)`:
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/border_router/dhcp6_pd_client.cpp:1268]: @gemini-code-assist the logic in the PR is correct and follows the RFC text exactly. The previous value of RT is used to determine the next RT according to the formula `RT = 2*RTprev + RAND*RTprev`:

> RT for each subsequent message transmission is based on the previous value of RT:
>
>       RT = 2*RTprev + RAND*RTprev

This maps to `(2 + RAND) * RTprev`, which is exactly what the code does, and not `(2 * RTprev) * (1 + RAND)`.

[the remark this answers is not in the document]
Comment (abtink) [reply src/core/border_router/dhcp6_pd_client.cpp:1268]: @gemini-code-assist, you are wrong about the `AddJitter()`. It returns (`mTimouet * (1 + RAND)`).

 Read its documentation which is already included in the PR. 

```c++
uint32_t Dhcp6PdClient::RetxTracker::AddJitter(uint32_t aValue, JitterMode aJitterMode)
{
    // Applies a random jitter to a given value based on the selected
    // jitter mode. For `kPositiveJitter`, the returned value is in
    // the range `[aValue, aValue * 1.1]`. For `kFullJitter`, the
    // range is `[aValue * 0.9, aValue * 1.1]`.
    //
    // Follows the RFC 8415 Section 15 rules regarding retransmission
    // timeout (RT) calculation, where the timeout includes a
    // randomization factor (RAND), which is a random number chosen
    // with a uniform distribution between -0.1 and +0.1, calculating
    // `RT + RAND * RT`.
    //
    // For a Solicit message's initial timeout, the randomization is
    // required to be positive, for which `kPositiveJitter` is used.

File: include/openthread/platform/infra_if.h [hunk 2]
```diff
@@ -182,6 +182,84 @@ otError otPlatGetInfraIfLinkLayerAddress(otInstance                    *aInstanc
...
+ * This function passes the ownership of @p aMessage to the platform layer. Platform MUST then free the message
+ * when no longer needed.
+ *
+ *
+ * @param[in] aInstance      The OpenThread instance.
+ * @param[in] aMessage       An `otMessage` containing the DHCPv6 payload. Ownership is passed to the platform layer.
+ * @param[in] aDestAddress   The IPv6 unicast destination address.
+ * @param[in] aInfraIfIndex  The index of the infrastructure interface from which to send the message.
+ */
+void otPlatInfraIfDhcp6PdClientSendUnicast(otInstance   *aInstance,
```
Comment (zwx1995esp) [inline include/openthread/platform/infra_if.h]: Hi, Abtink, I have a question:  is there a specific reason why we have separate functions for sending to unicast and multicast addresses? If the only difference between these two APIs is the destination address, could we consider merging them into a single API? Openthread stack passes the destination address, which might be a multicast address, to platform.

Comment (abtink) [reply include/openthread/platform/infra_if.h]: This is mainly intended to emphasize that the destination is a multicast address, serving as a hint to platform developers. This is particularly for platforms that may require specific behaviors for multicast transmissions (such as using a different type of socket, or certain network APIs).

Most platforms may not need this, in which case the platform implementation can simply have `SendMulticast()` call `SendUnicast()` with the multicast address.

Comment (zwx1995esp) [reply include/openthread/platform/infra_if.h]: Oh, Abtink, as you mentioned, `most platforms may not need this`. So how about we simply mention in the parameter description that the destination address might be a multicast address? Since most platforms are unlikely to implement two nearly identical functions.

Comment (abtink) [reply include/openthread/platform/infra_if.h]: It's okay either way; I can change it later.

By the way, you seem to be interested in this feature. Are you planning to use or deploy it? I will be adding testing (unit tests,etc) for this, but it would be great to also validate and test it on a physical device (deployments). Is this something that you might be able to help with?

Comment (zwx1995esp) [reply include/openthread/platform/infra_if.h]: Yes, Abtink, our platform uses LwIP stack. But the LwIP lacks support for stateful DHCPv6, so I'm vary glad to see that you provide support for this functionality. I can help with some testing based on the ESP platform, but I currently don't have a DHCPv6 server to assist with the testing. Do you have any suggestions on this?

Comment (zwx1995esp) [reply include/openthread/platform/infra_if.h]: Hi, Abtink, for DHCPv6 server, I think I can try with[ `kea-dhcp6`](https://kea.readthedocs.io/en/latest/arm/install.html) (It seems I can install it on my linux host and setup a test env). If this PR is ready, or almost ready, I can help with verifying this feature based the ESP platform.

Comment (abtink) [reply include/openthread/platform/infra_if.h]: Great. Thanks.

This is not yet ready (I'm still implementing unit tests, particularly to cover various edge cases). I will mark it as ready (by removing 'UNDER WORK' from the title and changing its status from 'draft') once ready.

Comment (abtink) [reply include/openthread/platform/infra_if.h]: I marked the PR now as "ready for review". No more changes planned for this PR itself

I will add a posix implementation of the newly added `otPlatInfraIfDhcp6*` APIs in a follow-up PR.

Comment (zwx1995esp) [reply include/openthread/platform/infra_if.h]: Ok, we will have a try with this feature.

File: include/openthread/platform/infra_if.h [hunk 3]
```diff
@@ -183,6 +184,66 @@ otError otPlatGetInfraIfLinkLayerAddress(otInstance                    *aInstanc
...
+ * @param[in] aDestAddress   The IPv6 multicast or unicast destination address.
+ * @param[in] aInfraIfIndex  The index of the infrastructure interface from which to send the message.
+ */
+void otPlatInfraIfDhcp6PdClientSend(otInstance   *aInstance,
+                                    otMessage    *aMessage,
+                                    otIp6Address *aDestAddress,
+                                    uint32_t      aInfraIfIndex);
+
+/**
+ * Callback from platform to notify the OpenThread stack of a received DHCPv6 message.
```
Comment (Vyrastas) [inline include/openthread/platform/infra_if.h]:
```suggestion
 * Callback from the platform to notify the OpenThread stack of a received DHCPv6 message.
```

File: include/openthread/platform/infra_if.h [hunk 4]
```diff
@@ -183,6 +184,66 @@ otError otPlatGetInfraIfLinkLayerAddress(otInstance                    *aInstanc
...
+ *
+ * This function is provided when `OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_CLIENT_ENABLE` is enabled.
+ *
+ * The platform calls this function whenever a DHCPv6 message is received on the client port (546) while the
```
Comment (Vyrastas) [inline include/openthread/platform/infra_if.h]:
```suggestion
 * The platform calls this function whenever a DHCPv6 message is received on the client port (546) while
```

File: src/core/border_router/dhcp6_pd_client.hpp [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    static constexpr uint16_t kInfiniteRetxCount     = 0;
+
+    static constexpr uint32_t kRetxDelayOnFailedTx = 330; // in msec
+
+    static constexpr uint32_t kIaid                = 0;
+    static constexpr uint8_t  kDesiredPrefixLength = 64;
+    static constexpr uint8_t  kDefaultPreference   = 0;
+    static constexpr uint32_t kJitterDivisor       = 10;
+
+    // The constants below are in unit of seconds
```
Comment (Vyrastas) [inline src/core/border_router/dhcp6_pd_client.hpp]:
```suggestion
    // The constants below are in seconds
```

Comment (abtink) [reply src/core/border_router/dhcp6_pd_client.hpp]: Thanks @Vyrastas. Updated in the new pushed commit.

**Review by Vyrastas — APPROVED** [review]
LGTM for docs

**Review by jwhui — APPROVED** [review]

File: include/openthread/platform/infra_if.h (hunk 2 above)
Comment (tanyanquan) [reply include/openthread/platform/infra_if.h]: Hi @abtink , we've been trying out this new feature over the past few days on the ESP platform and it's been working great :) just a small issue which we've faced and addressed in https://github.com/openthread/openthread/pull/11625

Thank you!

Comment (abtink) [reply include/openthread/platform/infra_if.h]: Great to hear! Thanks for your help validating it. 😃
